pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,476 @@
+func (r *RaftTracer) tryStore(tv *traceValue) (*traceValue, bool) {
+	if existingTv, loaded := r.m.LoadOrStore(tv.traced.Index, tv); loaded {
+		tv.logf(2, "duplicate registration ignored - existing trace: %s", existingTv)
+		existingTv.logf(2, "additional registration for same index: %s", tv)
+		r.destroy(tv)
+		return existingTv, false
+	}
+	return tv, true
+}
+
+// newTraceValue creates a new traceValue for the given traced entry. Note that
+// it doesn't pass `propCtx` as the first parameter since this isn't the
+// relevant context that should be used for logging and it can be nil.
+func (r *RaftTracer) newTraceValue(
+	te kvserverpb.TracedEntry, propCtx context.Context, propSpan *tracing.Span,
+) *traceValue {
+	tv := &traceValue{traced: te}
+	tv.ctx = logtags.AddTag(r.ctx, "id", redact.Safe(tv.String()))
+	tv.mu.seenMsgAppResp = make(map[raftpb.PeerID]bool)
+	tv.mu.propCtx = propCtx
+	tv.mu.propSpan = propSpan
+	return tv
+}
+
+// RegisterRemote registers a remote trace. This is called when we receive a
+// raft message over the wire with a request to continue tracing it.
+func (r *RaftTracer) RegisterRemote(te kvserverpb.TracedEntry) {
+	if !r.reserveSpace() {
+		return
+	}
+	// NB: We don't currently return remote traces, if we did, we would pass the
+	// remote ctx here and trace it.
+	if tv, created := r.tryStore(r.newTraceValue(te, nil, nil)); created {
+		tv.logf(1, "registering remote trace %s", tv)
+	}
+}
+
+// MaybeRegister is called on an entry that has been proposed to raft. This will
+// begin logging all subsequent updates to this entry. It returns true if the
+// registration is successful. A duplicate registration of the same index is
+// considered a success and returns true, however the older registration is kept
+// and this registration is ignored.
+func (r *RaftTracer) MaybeRegister(ctx context.Context, ent raftpb.Entry) bool {
+	// If the index is nil, then we can't trace this entry. This can happen if
+	// there is a leader/leaseholder spilt. We don't have an easy way to handle
+	// this today, so don't attempt to trace it.
+	if ent.Index == 0 {
+		log.VEvent(ctx, 2, "skip registering raft proposal without index")
+		return false
+	}
+
+	// Only register the entry if this is a traced context with verbose logging.
+	span := tracing.SpanFromContext(ctx)
+	if span == nil || span.RecordingType() != tracingpb.RecordingVerbose {
+		return false
+	}
+
+	// This must be the last conditional. If this returns true we must call
+	// storeEntryWithTracing to not leak a registered permit.
+	if !r.reserveSpace() {
+		log.VEvent(ctx, 2, "too many active raft traces, skipping")
+		return false
+	}
+
+	ctx, span = r.tracer.StartSpanCtx(ctx, "raft trace",
+		tracing.WithParent(span), tracing.WithFollowsFrom())
+	if tv, created := r.tryStore(
+		r.newTraceValue(kvserverpb.TracedEntry{
+			Index:   kvpb.RaftIndex(ent.Index),
+			TraceID: span.TraceID(),
+			SpanID:  span.SpanID(),
+		},
+			ctx,
+			span,
+		)); created {
```

Comment:
The formatting is a bit messed up. As an option:

```golang
if tv, created := r.tryStore(r.newTraceValue(kvserverpb.TracedEntry{
	Index:   kvpb.RaftIndex(ent.Index),
	TraceID: span.TraceID(),
	SpanID:  span.SpanID(),
}, ctx, span)); created {
	tv.logf(...)
}
```

Reply:
Changed